internal/vault_reader.go

```diff
@@ -0,0 +1,273 @@
+package internal
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	rabbitmqv1beta1 "github.com/rabbitmq/cluster-operator/api/v1beta1"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	vault "github.com/hashicorp/vault/api"
+)
+
+const defaultAuthPath string = "auth/kubernetes"
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . SecretReader
+type SecretReader interface {
+	ReadSecret(path string) (*vault.Secret, error)
+}
+
+type VaultSecretReader struct {
+	client *vault.Client
+}
+
+func (s VaultSecretReader) ReadSecret(path string) (*vault.Secret, error) {
+	secret, err := s.client.Logical().Read(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read Vault secret: %w", err)
+	}
+	return secret, nil
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . SecretStoreClient
+type SecretStoreClient interface {
+	ReadCredentials(path string) (CredentialsProvider, error)
+}
+
+type VaultClient struct {
+	Reader SecretReader
+}
+
+var ReadServiceAccountTokenFunc = ReadServiceAccountToken
+var ReadVaultClientSecretFunc = ReadVaultClientSecret
+var LoginToVaultFunc = LoginToVault
+
+var createSecretStoreClientOnce sync.Once
+var SecretClient SecretStoreClient
+var SecretClientCreationError error
+
+func GetSecretStoreClient(vaultSpec *rabbitmqv1beta1.VaultSpec) (SecretStoreClient, error) {
+	createSecretStoreClientOnce.Do(InitializeClient(vaultSpec))
+	return SecretClient, SecretClientCreationError
+}
+
+func InitializeClient(vaultSpec *rabbitmqv1beta1.VaultSpec) func() {
+	return func() {
+		// VAULT_ADDR environment variable will be the address that pod uses to communicate with Vault.
+		config := vault.DefaultConfig() // modify for more granular configuration
+		vaultClient, err := vault.NewClient(config)
+		if err != nil {
+			SecretClientCreationError = fmt.Errorf("unable to initialize Vault client: %w", err)
+			return
+		}
+
+		_, err = login(vaultClient, vaultSpec)
```

Comment:
This is a minor issue: When the client is first initialized, it looks like we are logging in twice. Once from `InitializeClient()` function so that we can fail fast if the credentials are wrong before scheduling the credentials renewal go routing. And then again, within the renewal go routing. 

If we wanted to fix it we would have to pass the first `vaultLoginResp` onto `renewToken` function and have an index, in the `for` loop, that tells us whether this is the first time and hence use the `vaultLoginResp` from the parameter list. Or instead ignore it and login. The code is certainly not as clean as the current one.

Reply:
Well spotted and apologies for not mentioning it earlier as it was something that caused me some concern. You are correct that on start up there are two Vault logins in very quick succession so we end up with a log that contains something like the following ...

```json
{"level":"info","ts":1637150459.7343986,"msg":"Authenticating to Vault"}
{"level":"info","ts":1637150459.746867,"msg":"Authenticating to Vault"}
```

To avoid this I also looked at keeping the `login()` call out of `InitializeClient()` altgether and only carrying it out in the renewal thread but that turned worse IMHO since it created a tiny window of time where the Vault client was trying to carry out its work without a valid Vault token and the log output contained a smattering of "unable to retrieve credentials" messages which looked worse. In the interests of trying to get a strawman in front of you sooner rather than later I opted for the former approach since it is not as messy in the logs and only happens on the first reconciliation after startup. I will see if I can take another pass at this.

Reply:
@MarcialRosales I've just pushed a small change to the code which stops the double logins occurring on startup. The `InitializeClient()` function no longer attempts a Vault login but instead [waits](https://github.com/georgeharley/messaging-topology-operator/blob/239-vault-cluster-authentication/internal/vault_reader.go#L73-L74) for the `renewToken()` function to notify it of the outcome of the its login attempt. The notification comes via a buffered channel passed to `renewToken()`.

`renewToken()` only writes to the channel [on the first login attempt](https://github.com/georgeharley/messaging-topology-operator/blob/239-vault-cluster-authentication/internal/vault_reader.go#L205-L214). If that first attempt fails then it seemed like a good opportunity to fail fast and bail out of proceeding with the whole token renewal loop. 

Have tried this on a couple of K8s clusters and now only seeing the one "Authenticating to Vault" log message on startup. In the unhappy scenario where the intial Vault login attempt fails then the "Lifecycle management of Vault token will not be carried out" message gets logged and all subsequent reconcile attempts fail with the Vault error.